test/AutoRest.TestServer.Tests/Mgmt/TestProjects/TestProjectTests.cs

```diff
@@ -285,5 +290,74 @@ public void ValidateSubscriptionExtensionsListResourceByName()
+            var library = output.Context.Library;
+            foreach (var model in library.Models)
+            {
+                if (model is MgmtObjectType mgmtObject)
+                    ValidateModelRequiredCtorParams(mgmtObject.ObjectSchema);
+            }
+            foreach (var resourceData in library.ResourceData)
+            {
+                ValidateModelRequiredCtorParams(resourceData.ObjectSchema);
+            }
+        }
+
+        private void ValidateModelRequiredCtorParams(ObjectSchema objectSchema)
+        {
+            var requiredParams = objectSchema.Properties.Where(p => p.Schema is not ConstantSchema && p.Required.HasValue && p.Required.Value);
+
+            Type generatedModel = GetType(objectSchema.Name + "Data") ?? GetType(objectSchema.Name);
+            Assert.NotNull(generatedModel, $"Generated type not found for {objectSchema.Name}");
+            ConstructorInfo leastParamCtor = GetLeastParamCtor(generatedModel);
+            ConstructorInfo baseLeastParamCtor = GetLeastParamCtor(generatedModel.BaseType);
+            var fullRequiredParams = requiredParams.Select(p => p.SerializedName).Concat(baseLeastParamCtor?.GetParameters().Select(p => p.Name)).Distinct();
+            Assert.NotNull(leastParamCtor, $"Ctor not found for {objectSchema.Name}");
+            Assert.AreEqual(fullRequiredParams.Count(), leastParamCtor.GetParameters().Length, $"{objectSchema.Name} had a mismatch in required ctor params");
+            if (fullRequiredParams.Count() != 0)
+            {
+                foreach (var param in fullRequiredParams)
+                {
+                    Assert.NotNull(leastParamCtor.GetParameters().FirstOrDefault(p => string.Equals(p.Name, param, StringComparison.InvariantCultureIgnoreCase)), $"{param} was not found in {objectSchema.Name}'s ctor");
+                }
+            }
+        }
+
+        private ConstructorInfo GetLeastParamCtor(Type generatedModel)
+        {
+            ConstructorInfo leastParamCtor = null;
+
+            if (generatedModel == null)
+                return leastParamCtor;
+
+            foreach (var ctor in generatedModel.GetConstructors(BindingFlags.Instance | BindingFlags.NonPublic | BindingFlags.Public))
+            {
+                if (ctor.GetParameters().Length < (leastParamCtor == null ? int.MaxValue : leastParamCtor.GetParameters().Length))
+                    leastParamCtor = ctor;
+            }
+            return leastParamCtor;
+        }
+
+        protected void ValidatePublicCtor(Type model, string[] paramNames, Type[] paramTypes)
+        {
+            var ctors = model.GetConstructors(System.Reflection.BindingFlags.Public | System.Reflection.BindingFlags.Instance);
+            Assert.AreEqual(1, ctors.Length);
+            var ctor = ctors.First();
+            var parameters = ctor.GetParameters();
+            for (int i = 0; i < parameters.Length; i++)
```

Comment:
Maybe this could be a foreach?

Reply:
need to index multiple arrays with the same indexer